src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs

```diff
@@ -1596,10 +1596,10 @@ public override bool HandleCall (MethodBody callingMethodBody, MethodReference c
 								var requiredMemberTypes = GetDynamicallyAccessedMemberTypesFromBindingFlagsForConstructors (bindingFlags);
 
 								// Special case the public parameterless constructor if we know that there are 0 args passed in
-								if (ctorParameterCount == 0 &&
-									requiredMemberTypes.HasFlag (DynamicallyAccessedMemberTypes.PublicConstructors) &&
-									!requiredMemberTypes.HasFlag (DynamicallyAccessedMemberTypes.NonPublicConstructors))
-									requiredMemberTypes = DynamicallyAccessedMemberTypes.PublicParameterlessConstructor;
+								if (ctorParameterCount == 0 && requiredMemberTypes.HasFlag (DynamicallyAccessedMemberTypes.PublicConstructors)) {
+									requiredMemberTypes &= ~DynamicallyAccessedMemberTypes.PublicConstructors;
+									requiredMemberTypes |= DynamicallyAccessedMemberTypes.PublicParameterlessConstructor;
```

Comment:
I must be missing something :-(

I think this is basically a no-op. `PublicConstructors = 3`, `PublicParameterlessConstructor = 1`, so `HasFlags(PublicConstructors)` should imply `PublicParameterlessConstructor`.
I wonder how come the original test doesn't fail with this change. The test `TestNullArgsOnAnnotatedType` should not work with this change if I understand it correctly.

Reply:
Sorry, I didn't describe the problem very clearly. This is to address the case where you use Public | NonPublic binding flags. The special case before this change will kick in only when the requirements have Public & !NonPublic - so they wouldn't apply. But we still don't need to keep all public ctors (we know there's only one parameter) - the Public requirement should be downgraded to PublicParameterless, regardless of whether NonPublic ctors were required.

`TestNullArgsOnAnnotatedType` will still hit this special case because it doesn't use NonPublic binding flags.

For more context, we started getting this warning:
```
/_/src/libraries/System.Data.Common/src/System/Data/DataSet.cs(1100,13): error IL2072: (NETCORE_ENGINEERING_TELEMETRY=Build) System.Data.DataSet.CreateInstanceOfThisType(): '#0' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors' in call to 'System.Object System.Activator::CreateInstance(System.Type,System.Boolean)'. The return value of method 'System.Type System.Object::GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

This is for the CreateInstance call here:

Reply:
I was reading this wrong. I get it now - sorry - no idea what I was thinking before.